Count all usage from failures to unmarshal and validate json

ericvanlare opened the pull request:
* Added tallying of usage for retries
* Lots of nil response returns, I replaced these (except for Cohere, stubbed that as a TODO) with empty responses containing only usage data

h0rv commented on pkg/instructor/chat.go:
```diff
@@ -48,12 +62,66 @@ func chatHandler(i Instructor, ctx context.Context, request interface{}, respons
+	return i.EmptyResponseWithUsage(usage), errors.New("hit max retry attempts")
+}
+
+func nilChatRespWithUsage(response interface{}) interface{} {
```
we should add the usage method as an abstract method to be implemented in each provider, so we do not need to worry about provider specific logic in this file

h0rv commented on pkg/instructor/chat.go:
```diff
@@ -48,12 +62,66 @@ func chatHandler(i Instructor, ctx context.Context, request interface{}, respons
+	return response
+}
+
+func countUsageFromResp(response interface{}, usage *UsageSum) {
```
add as abstract method to interface

h0rv commented on pkg/instructor/chat.go:
```diff
@@ -48,12 +62,66 @@ func chatHandler(i Instructor, ctx context.Context, request interface{}, respons
+	}
+}
+
+func addUsageSumToResp(response interface{}, usage *UsageSum) interface{} {
```
add as abstract method to interface

h0rv commented:
Thanks for the PR - this is a good feature! It looks good, I just requested a change to add an extra abstract method or 2, so we don't need to import each LLM provider into the main `instructor` interface logic

ericvanlare commented:
Thanks for the feedback @h0rv , good call, I moved all the provider-specific logic into their respective chat files and added abstract methods for them in `Instructor`

h0rv reviewed (approved):
Looks great - thanks for the PR!
